### PR TITLE
fix(ESM): Add operators/package.json for ESM support, fixes #3227

### LINKED ISSUE #3227: Should libraries that depend on rxjs use path-mappings/tree shaking?
👋  This might at first glance seem like an obvious question, but let me explain:

By default in v6 if you import anything it's going to resolve the files from the project root. e.g. `import { Observable } from 'rxjs'` is going to come from `node_modules/rxjs/Rx.js` which re-exports it from `node_modules/rxjs/internal/Observable.js`.

If a library uses the [path-mappings and tree shaking](https://github.com/ReactiveX/rxjs/blob/master/doc/pipeable-operators.md#build-and-treeshaking) they would use custom resolve overrides. So in that case, Observable would come from `node_modules/rxjs/_esm5/index.js` which re-exports it from `node_modules/rxjs/_esm5/internal/Observable.js`. Note the fact that this is inside `rxjs/_esm5`.

Then let's say that someone imports my library (e.g. https://redux-observable.js.org) which had done this, but **the consumer does not set up their own webpack resolve overrides** so they'll get Observable from `node_modules/rxjs/internal/Observable.js` and now we have two copies of rxjs in the same project. `Observable1 instanceof Observable2 === false`. If they use pipeable operators they probably won't immediately run into errors, if ever, but their bundle size will be silently larger and might later get cases where `Observable1 instanceof Observable2 === false`, which is quite the hard thing to diagnose for most.

If the library doesn't use path-mappings, the same problem exists in reverse; if the library consumer do use it, they'll again get their own copy of RxJS.

***

This was brought up previously when we had rxjs-es (because a redux-observable user hit it) and discussed including the esm5 versions, but I don't recall a solution ever being found.

Yes, this sucks. AFAIK the only real solution would be have the esm5 version in its own package (or make it the default package and put the other stuff in another). Interestingly, if that was done I think people wouldn't need to have path-mappings and instead could just alias `"rxjs": require.resolve('rxjs-esm5')` or whatever it would be called. This could still give you two copies but to me feels like it reduces the likelihood as long as libs set their peerDep correctly.

🔥

### PR DESCRIPTION
**Pull request by jayphelps**
See #3227

I didn't do anything for `@reactivex/rxjs` package because I'm hopeful for its elimination (#2916) and I'm not entirely sure how it works lol.

### CODE REVIEW

File: .make-packages.js [hunk 1]
```diff
@@ -40,7 +40,9 @@ fs.removeSync(PKG_ROOT);
 let rootPackageJson = Object.assign({}, pkg, {
   name: 'rxjs',
   main: './index.js',
-  typings: './index.d.ts'
+  typings: './index.d.ts',
+  module: './_esm5/index.js',
+  es2015: './_esm2015/index.js'
```
Comment (jayphelps) [inline .make-packages.js]: It's hard to keep everything straight, but AFAICT this also makes the whole path-mappings.js thing unnecessary if you're using a newish version of webpack or rollup.

Would be cool to still include instructions (and maybe path-mappings still) for people who are stuck on older build pipelines.

Comment (simonbuchan) [reply .make-packages.js]: Yes, if users only import from `"rxjs"` and `"rxjs/operators"`, importing `"rxjs/index"` or (more realistically) `"rxjs/Rx"` would not be mapped.

**Comment by simonbuchan** [comment]
This says "fixes #3227", but it doesn't handle webpack native tree-shaking? I'd have to check the quality difference between having `"sideEffects": false` and not in webpack 4, I think with the new uglifyjs in there too it should be able to optimize without it, but it might be slower or something.

**Comment by jayphelps** [comment]
>  it doesn't handle webpack native tree-shaking?

Can you clarify?

**Comment by simonbuchan** [comment]
See [the example](https://github.com/webpack/webpack/tree/f6e366b4be1cfe2770251a890d93081824789209/examples/side-effects)
It's why I called out "webpack native" tree shaking and mentioned uglify should still work - but presumably the option exists for a reason.

**Comment by jayphelps** [comment]
@simonbuchan oh yeah `sideEffects: false` might be able to be supported but that's being tracked in #3212 not #3227. Perhaps confusing the two? 😄

**Comment by jasonaden** [comment]
Agreed on #3212 vs #3227. This PR should close #3227 (as long as a `package.json` gets added for other secondary entry points such as `ajax, websocket`, and `testing`).

#3212 will have another PR shortly (I've got that one).

**Comment by simonbuchan** [comment]
@jayphelps Probably! But #3227 reads to me an umbrella issue of "make it just work™".

@jasonaden Is `"rxjs/Rx"` going away too?

**Comment by jayphelps** [comment]
@jasonaden 

> I think we also need a similar package.json file in ajax, testing, and websocket, right?

Ohhh right! Thanks for catching that!

@simonbuchan sorry for the confusion! My primary intention with that ticket was how to prevent two copies of RxJS. 🕺

**Comment by jayphelps** [comment]
@jasonaden I've updated to include package.json's for ajax/websocket/testing

**Pull request merged**